update shell node version to 6.3.1

zaggino opened the pull request:
ref: https://github.com/adobe/brackets/issues/11748

I've tried doing this, but I get the following error which is not comprehensible for me:
```
>> Error: Command failed: C:\WINDOWS\system32\cmd.exe /s /c "bash scripts/fix-msvc.sh"
>>       2 [main] bash 3416 C:\Program Files\Git\usr\bin\bash.exe: * fatal error in forked process - fork: can't reserve memory for parent stack 0x600000 - 0x800000, (child has 0x400000 - 0x600000), Win32 error 487
>>     513 [main] bash 3416 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump
>>       2 [main] bash 8080 fork: child -1 - forked process 3416 died unexpectedly, retry 0, exit code 0x100, errno 11
>> scripts/fix-msvc.sh: fork: retry: No child processes
>> 1004878 [main] bash 10932 C:\Program Files\Git\usr\bin\bash.exe: * fatal error in forked process - fork: can't reserve memory for parent stack 0x600000 - 0x800000, (child has 0x400000 - 0x600000), Win32 error 487
>> 1005294 [main] bash 10932 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump
>> 1100847 [main] bash 8080 fork: child -1 - forked process 10932 died unexpectedly, retry 0, exit code 0x100, errno 11
>> scripts/fix-msvc.sh: fork: retry: No child processes
>> 3108069 [main] bash 8804 C:\Program Files\Git\usr\bin\bash.exe: * fatal error in forked process - fork: can't reserve memory for parent stack 0x600000 - 0x800000, (child has 0x400000 - 0x600000), Win32 error 487
>> 3108730 [main] bash 8804 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump
>> 3207844 [main] bash 8080 fork: child -1 - forked process 8804 died unexpectedly, retry 0, exit code 0x100, errno 11
>> scripts/fix-msvc.sh: fork: retry: No child processes
>> 7224872 [main] bash 11240 C:\Program Files\Git\usr\bin\bash.exe: * fatal error in forked process - fork: can't reserve memory for parent stack 0x600000 - 0x800000, (child has 0x400000 - 0x600000), Win32 error 487
>> 7226544 [main] bash 11240 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump
>> 7316396 [main] bash 8080 fork: child -1 - forked process 11240 died unexpectedly, retry 0, exit code 0x100, errno 11
>> scripts/fix-msvc.sh: fork: retry: No child processes
>> 15320804 [main] bash 11272 C:\Program Files\Git\usr\bin\bash.exe: * fatal error in forked process - fork: can't reserve memory for parent stack 0x600000 - 0x800000, (child has 0x400000 - 0x600000), Win32 error 487
>> 15321198 [main] bash 11272 cygwin_exception::open_stackdumpfile: Dumping stack trace to bash.exe.stackdump
>> 15409163 [main] bash 8080 fork: child -1 - forked process 11272 died unexpectedly, retry 0, exit code 0x100, errno 11
```
also, why do we need npm?

ingorichter commented:
I don't exactly see why we need npm. Perhaps @jasonsanjose recalls why we need it.

zaggino commented:
btw npm is only part of `node-win` so it makes me wonder why there is `npm` for windows and not for other envs

ingorichter commented:
That's a good question to ask. What happens if you don't download it? Are you able to build and run the tests?

jasonsanjose commented:
Whew. That took some digging but there's the answer https://github.com/adobe/brackets-shell/commit/c913ec882c24051f17495490a91296f0fc7850ba. We used npm to install a dependency on `ws` for websocket support. 

In https://github.com/adobe/brackets-shell/pull/209 we moved that code from @joelrbrandt's git repo into the brackets-shell repo, but we never removed the npm dependency.

Looks safe to remove to me.

ingorichter commented:
Great! Thanks for digging this up @jasonsanjose. @zaggino do want to remove it in this PR?

zaggino commented:
Yes, because npm package is no longer on the nodejs server and the format of the tar in npm registry is a bit different.

ficristo commented:
@zaggino @ingorichter Is this blocked on something?

abose commented:
@ficristo Have you tested if 
1. the existing file watching mechanism is properly woking with updated node engine?
2. installing extensions from the extension manager works.
3. Unit tests relating to extension pass
4. instant search is working
5. Unit tests relating to instant search is working.
Once all these are verified, I could run some test cases with updated node.

abose commented:
Oh wait! it was @zaggino 's PR :sweat_smile: Wrongly addressed to @ficristo .

zaggino commented:
This is not finished, I've run into `fatal error in forked process` and wasn't able to get around it. Compiling brackets works fine on my machine but not with this PR.

zaggino commented:
if anyone wants to try to compile the shell with this PR, go ahead ... maybe something is wrong with my machine, I'll test this on Mac when I get a chance

ficristo commented on tasks/setup.js:
```diff
@@ -377,22 +377,12 @@ module.exports = function (grunt) {
 
         var done        = this.async(),
```
Now this is unused: remove or call it.
(I've had some trouble testing the PR because of this)

ficristo commented on tasks/setup.js:
```diff
@@ -377,22 +377,12 @@ module.exports = function (grunt) {
-
-        // unzip NPM
-        unzip(npmFile, "deps/node").then(function () {
-            nodeWriteVersion();
```
Why it was removed? Is not needed anymore?

zaggino replied:
Read the comments in the PR, jasonsanjose commented on 18 Dec 2015

ficristo replied:
@zaggino I think the call to `nodeWriteVersion` should be left.

ficristo commented:
Some days ago I tryed this PR and had some problems.
I tryed again today and seems to run fine on Windows... :confused:

ingorichter commented:
I tried this today and it failed to download the node binary. The current download URL is `http://nodejs.org/dist/v4.4.4/node-v4.4.4-darwin-x86.tar.gz`, but this version is no longer available. Only the `x64` version is available for OSX. So, this needs to be updated and then the question is: does it compile and work with the x64 version?

zaggino commented:
I'd be great if someone could tackle this...
https://github.com/zaggino/brackets-eslint/issues/36

AdriVanHoudt commented:
what is blocking this?

ficristo commented:
@AdriVanHoudt first things that come to my mind:
- I left some comments that should be answered.
- I think the version for OSX should be changed. See @ingorichter comment.
- Verify all steps mentioned by @abose.
- What we will do for Linux since is building from another branch?

ingorichter commented:
I tried again last night and we have to change to the x64 variant. Changing it to x64 makes the download and build succeed. I haven't tested yet if Brackets launches and behaves properly (my machine had some thermal issues and shutdown down :-/). I can test it tonight on a different machine.

zaggino commented:
I've changed Mac to the 64 bits; I've updated to the latest node (as there's little point to upgrading to v4 now that v6 is out); Removed some remaining npm references;

I'd love if someone from Adobe could try to build this on the machines that they use to build releases as my current Windows machine stopped being able to compile brackets-shell for whatever reason.

If I have a valid build of shell, I can start working on a PR of making https://github.com/adobe/brackets work with it.

zaggino commented:
Guys, what about not porting fsevents_win and using https://github.com/paulmillr/chokidar for watching file system changes? If it's good for others, it should be good for Brackets, no? @ficristo @ingorichter

ficristo commented:
@zaggino I'm totally in favor: https://github.com/adobe/brackets/pull/12190

zaggino commented:
Ok, I'm gonna try building this shell and run tests using https://github.com/adobe/brackets/pull/12190 branch, we'll see what happens. Anyone else is welcome to do the same :)

ingorichter commented:
@zaggino I like the idea. Frees some resources for us. I'll give it a try tonight

zaggino commented:
Now that https://github.com/adobe/brackets/pull/12647 has landed in master, there're no other native dependencies that would block this from merging.

Please review/merge @ingorichter @ficristo 

I'd love to see this land for 1.8 to solve https://github.com/zaggino/brackets-eslint/issues/36 and also https://github.com/adobe/brackets/issues/11748#issuecomment-228580962

ficristo commented:
To test this I've run only the test suite.
I see some consistent failures:
- Integration \ FindInFiles
- MainView \ MainViewManager
- Extensions \ Code Folding
- Extensions \ HealthData

but I doubt any of these is related to this PR.
For example I'm sure the FindInFiles is on master too, but I didn't test master to check the others.

ficristo commented:
I tryed both on Windows10 and OSX and see some consistent failures on `Integration \ FindInFiles` and `Extensions \ HealthData` but I'm sure they are on master too.
@zaggino have you checked @abose [comment](https://github.com/adobe/brackets-shell/pull/543#issuecomment-219078895)?
Otherwise I think this is ready.

zaggino commented:
@ficristo @abose 

1. yes, file watching was updated to support node 6 and that's already in master
2. yes, installing extensions from the extension manager works.
3. yes, unit tests relating to extension manager pass
4. yes, instant search is working
5. yes, unit tests related to instant search are working

running this branch and latest master, no unit test is failing, from integration tests there are two failures which are definitely not related to this

nethip commented:
@zaggino and everyone else on this thread. Thanks for your amazing efforts in taking up this herculean task. As always, I am in awe :+1:

nethip commented on Gruntfile.js:
```diff
@@ -71,7 +71,7 @@ module.exports = function (grunt) {
             "node-mac": {
                 "dest"      : "<%= downloads %>",
-                "src"       : "http://nodejs.org/dist/v<%= node.version %>/node-v<%= node.version %>-darwin-x86.tar.gz"
+                "src"       : "http://nodejs.org/dist/v<%= node.version %>/node-v<%= node.version %>-darwin-x64.tar.gz"
```
We might want to leave the mac node to be 32 bit, as our native node extensions are still 32 bit and they would not load on 64 bit node.

zaggino replied:
Which native node extensions are you talking about?

zaggino replied:
Node doesn't have 32-bit builds for Mac anymore:
![image](https://cloud.githubusercontent.com/assets/1067319/18193072/3fc6dd14-711d-11e6-9249-5e9580c779be.png)

nethip replied:
Yeah. Now that we have moved to Chokidar, I don't think we will have any references to native node extensions. I was referring to this library which we were using earlier for file watching.

https://github.com/adobe/brackets/tree/release/src/filesystem/impls/appshell/node/node_modules/fsevents

I think then it should be fine to go with 64 bit.
